source/common/http/filter_manager.cc

```diff
@@ -902,16 +929,28 @@ void FilterManager::encodeHeaders(ActiveStreamEncoderFilter* filter, ResponseHea
     (*entry)->end_stream_ = state_.encoding_headers_only_ ||
                             (end_stream && continue_data_entry == encoder_filters_.end());
     FilterHeadersStatus status = (*entry)->handle_->encodeHeaders(headers, (*entry)->end_stream_);
-    if ((*entry)->end_stream_) {
-      (*entry)->handle_->encodeComplete();
-    }
+
+    ASSERT(!(status == FilterHeadersStatus::ContinueAndEndStream && (*entry)->end_stream_),
+           "Filters should not return FilterHeadersStatus::ContinueAndEndStream from encodeHeaders "
+           "when end_stream is already true");
+    ASSERT(!(status == FilterHeadersStatus::ContinueAndDontEndStream && !(*entry)->end_stream_),
+           "Filters should not return FilterHeadersStatus::ContinueAndDontEndStream from "
+           "encodeHeaders when end_stream is already false");
+
     state_.filter_call_state_ &= ~FilterCallState::EncodeHeaders;
     ENVOY_STREAM_LOG(trace, "encode headers called: filter={} status={}", *this,
                      static_cast<const void*>((*entry).get()), static_cast<uint64_t>(status));
 
     (*entry)->encode_headers_called_ = true;
-    const auto continue_iteration =
-        (*entry)->commonHandleAfterHeadersCallback(status, state_.encoding_headers_only_);
+
+    // encoding_headers_only_ is set if the filter returns ContinueAndEndStream.
+    const auto continue_iteration = (*entry)->commonHandleAfterHeadersCallback(
+        status, end_stream, state_.encoding_headers_only_);
+
+    // If this filter ended the stream, encodeComplete() should be called for it.
+    if ((*entry)->end_stream_ || state_.encoding_headers_only_) {
+      (*entry)->handle_->encodeComplete();
+    }
```

Comment:
Is this change related to making the ContinueAndDontEndStream path work? Seems unrelated but maybe I'm missing something

Reply:
This is unrelated. It is a small bug I found while working on this (refer to the above conversation with Matt). It is actually related to the `ContinueAndEndStream` path. This makes sure that `encodeComplete()` is called for the filter that returns `ContinueAndEndStream`.